src/D2L.CodeStyle.Analyzers/Async/Generator/SyncGenerator.cs

```diff
@@ -0,0 +1,182 @@
+		IncrementalValuesProvider<MethodGenerationResult> generatedMethods =
+			methodsToGenerate
+				.Select( GenerateSyncMethod )
+				// Filter out things we simply couldn't generate anything for.
+				.Where( static mr => mr.HasValue )
+				.Select( static ( mr, _ ) => mr!.Value );
+
+		// Collect and generate the output files
+		//
+		// Opportunity for better caching here -- changing one source file
+		// should only impact one output file. Currently we only generate the
+		// methods when their source file changed, but we re-collect all of
+		// that when any file with [GenerateSync] has changed.
+		IncrementalValuesProvider<FileGenerationResult> generatedFiles =
+			generatedMethods.Collect()
+				.SelectMany( GenerateFiles );
+
+		context.RegisterSourceOutput( generatedFiles, WriteFiles );
+	}
+
+	private static bool IsInterestingLookingMethod(
+		SyntaxNode node,
+		CancellationToken _
+	) {
+		if( node is not MethodDeclarationSyntax method ) {
+			return false;
+		}
+
+		// We only care about async methods
+		if( !method.Modifiers.Any( t => t.IsKind( SyntaxKind.AsyncKeyword ) ) ) {
+			return false;
+		}
+
+		// The ones we care about have to have a particular attribute -- so
+		// perk up when we see _any_ attributes (we don't have a SemanticModel
+		// in this method.)
+		return method.AttributeLists.Any();
+	}
+
+	private static MethodDeclarationSyntax? ExtractSyntax(
+		GeneratorSyntaxContext context,
+		CancellationToken cancellationToken
+	) {
+		var methodDeclaration = (MethodDeclarationSyntax)context.Node;
+
+		var attributes = methodDeclaration.AttributeLists
+			.SelectMany( al => al.Attributes );
+
+		foreach( var attribute in attributes ) {
+			// We have the semantic model now, so make sure this method has
+			// the right attribute.
+			if( IsGenerateSyncAttribute( context.SemanticModel, attribute, cancellationToken ) ) {
+				return methodDeclaration;
+			}
+		}
+
+		return null;
+	}
+
+	internal static bool IsGenerateSyncAttribute(
+		SemanticModel model,
+		AttributeSyntax attribute,
+		CancellationToken cancellationToken
+	) {
+		var attributeConstructorSymbol = model.GetSymbolInfo( attribute, cancellationToken ).Symbol as IMethodSymbol;
+
+		if( attributeConstructorSymbol == null ) {
+			return false;
+		}
+
+		return attributeConstructorSymbol.ContainingType.ToDisplayString()
+			== "D2L.CodeStyle.Annotations.GenerateSyncAttribute";
+	}
+
+	private static MethodGenerationResult? GenerateSyncMethod(
+		(MethodDeclarationSyntax MethodDeclaration, Compilation Compilation) data,
+		CancellationToken cancellationToken
+	) {
+		if( !data.Compilation.ContainsSyntaxTree( data.MethodDeclaration.SyntaxTree ) ) {
+			return null;
+		}
+
+		var model = data.Compilation.GetSemanticModel( data.MethodDeclaration.SyntaxTree );
+		var methodSymbol = model.GetDeclaredSymbol( data.MethodDeclaration, cancellationToken );
+
+		if( methodSymbol == null || methodSymbol.Kind == SymbolKind.ErrorType ) {
+			return null;
+		}

```

Comment:
Are these all guards for "the compilation is otherwise broken, so ignore"? Or are there other scenarios where we might expect these to happen?

Reply:
I think only compilation broken, and honestly I'm not even sure in what scenarios it would happen there. `null` is possible by the contract, `ErrorType` honestly I sprinkle around when I remember to (that one definitely means "compile is broken but we're still running things anyway" and its safe to do whatever basically.)